File: examples/run_server.py
import os
import os.path as osp
import time
os.environ["GREEN_BACKEND"] = "gevent"
from greenthread.monkey import monkey_patch; monkey_patch()

from mqsrv.server import make_server, run_server
from mqsrv.exc import BaseException as BE

def foo():
    print("foo")

def add(x, y):
    print(f"add: {x}+{y}")
    return x + y

def test_str(s):
    print(s)
    return "test_str"+s

def test_vec(lst):
    print(lst)
    return [1, 2, 3, 4, 5]

def test_map(d):
    print(d)
    return {"a": 1, "b": 2, "c": 3}

def test_err():
    raise BE("test_err")

class Test:
    def __init__(self) -> None:
        pass

    def sub(self, a, b):
        print(f"sub: {a}-{b}")
        return a - b
    
def main():
    server = make_server(
        rpc_routing_key="test_rpc_queue"
    )
    t = Test()
    server.register_rpc(add, "add")
    server.register_rpc(foo, "foo")
    server.register_rpc(t.sub, "sub")
    server.register_rpc(test_str, "test_str")
    server.register_rpc(test_vec, "test_vec")
    server.register_rpc(test_map, "test_map")
    server.register_rpc(test_err, "test_err")
    run_server(server)
    
if __name__ == "__main__":
    main()

File: examples/run_server.cc
#include "server.h"

using namespace ccmqsrv;

void foo()
{
  std::cout << "foo" << std::endl;
}

int add(int a, int b)
{
  spdlog::info("add({}, {})", a, b);
  // sleep(2);
  return a+b;
}

std::string test_str(const std::string& str)
{
  spdlog::info("test_str({})", str);
  return "test_str: " + str;
}

std::vector<int> test_vec(std::vector<int> vec)
{
  spdlog::info("test_vec");
  for (auto& i : vec) {
    spdlog::info("{}", i);
  }
  std::vector<int> res{1, 2, 3, 4, 5};
  return res;
}

std::map<std::string, int> test_map(std::map<std::string, int> m)
{
  spdlog::info("test_map");
  for (auto& [key, value] : m) {
    spdlog::info("{} : {}", key, value);
  }
  std::map<std::string, int> res{{"a", 1}, {"b", 2}, {"c", 3}};   
  return res;
}

void test_err()
{
  throw std::runtime_error("test_err");
}

class Test {
public:
  int sub(int a, int b)
  {
    spdlog::info("sub({}, {})", a, b);
    return a-b;
  }
};

int main()
{
  spdlog::set_level(spdlog::level::debug);
  auto server = make_server("", "test_rpc_queue");
  Test test;
  server.register_rpc(&Test::sub, &test, "sub");
  server.register_rpc(add, "add");
  server.register_rpc(foo, "foo");
  server.register_rpc(test_str, "test_str");
  server.register_rpc(test_vec, "test_vec");
  server.register_rpc(test_map, "test_map");
  server.register_rpc(test_err, "test_err");
  run_server(server);
  return 0;
}